### PR TITLE
Server: support `GetPrevRegion`

### PR DESCRIPTION
**Pull request by liukun4515**
## What have you changed? (required)

Support the `GetPrevRegion`

## What are the type of the changes (required)?

- New feature (non-breaking change which adds functionality)

## How has this PR been tested (required)?
UT
## Refer to a related PR or issue link (optional)
#1055 
[kvproto](https://github.com/pingcap/kvproto/pull/289)

### CODE REVIEW

**Comment by liukun4515** [comment]
PTAL @disksing @nolouch

File: pd-client/client.go [hunk 1]
```diff
@@ -598,6 +600,30 @@ func (c *client) GetRegion(ctx context.Context, key []byte) (*metapb.Region, *me
 	return resp.GetRegion(), resp.GetLeader(), nil
 }
 
+func (c *client) GetPrevRegion(ctx context.Context, key []byte) (*metapb.Region, *metapb.Peer, error) {
+	if span := opentracing.SpanFromContext(ctx); span != nil {
+		span = opentracing.StartSpan("pdclient.GetPrevRegion", opentracing.ChildOf(span.Context()))
+		defer span.Finish()
+	}
+	start := time.Now()
+	defer func() { cmdDuration.WithLabelValues("get_prev_region").Observe(time.Since(start).Seconds()) }()
```
Comment (siddontang) [inline pd-client/client.go]: do we really need these two cmdDuration and requestDuration metrics now?
@disksing

File: pd-client/client_test.go [hunk 2]
```diff
@@ -175,6 +176,45 @@ func (s *testClientSuite) TestGetRegion(c *C) {
...
+			EndKey:   []byte{byte(i + 1)},
+			Peers:    []*metapb.Peer{peer},
+		}
+		regions = append(regions, r)
+		req := &pdpb.RegionHeartbeatRequest{
+			Header: newHeader(s.srv),
+			Region: r,
+			Leader: peer,
+		}
+		logrus.Info("client:", req)
```
Comment (disksing) [inline pd-client/client_test.go]: The log can be removed.

File: server/core/region_tree.go [hunk 3]
```diff
@@ -115,6 +115,24 @@ func (t *regionTree) search(regionKey []byte) *metapb.Region {
...
 
+// searchPrev returns the previews region of the region where the regionKey is located.
+func (t *regionTree) searchPrev(regionKey []byte) *metapb.Region {
+	curRegion := &metapb.Region{StartKey: regionKey}
+	curRegionItem := t.find(curRegion)
+	if curRegionItem == nil {
+		return nil
+	}
+	prevRegionItem, _ := t.getAdjacentRegions(curRegionItem.region)
+	// previews region may not exist?
```
Comment (disksing) [inline server/core/region_tree.go]: That's right. You can remove the question mark.

File: pd-client/client.go (hunk 1 above)
Comment (siddontang) [reply pd-client/client.go]: so I think we can just one one for other commands.

Comment (disksing) [reply pd-client/client.go]: I have no objection of recording only cmdDuration for commands, except TSO. /cc @liukun4515

File: pd-client/client.go [hunk 4]
```diff
@@ -47,6 +47,8 @@ type Client interface {
 	// Also it may return nil if PD finds no Region for the key temporarily,
 	// client should retry later.
 	GetRegion(ctx context.Context, key []byte) (*metapb.Region, *metapb.Peer, error)
+	// GetPrevRegion gets the previews region and its leader Peer of the region where the key is located.
```
Comment (siddontang) [inline pd-client/client.go]: /previews/previous/

File: server/core/region.go [hunk 5]
```diff
@@ -561,6 +561,15 @@ func (r *RegionsInfo) SearchRegion(regionKey []byte) *RegionInfo {
 	return r.GetRegion(region.GetId())
 }
 
+// SearchPrevRegion search previews RegionInfo from regionTree
```
Comment (rleungx) [inline server/core/region.go]: s/search/searches

File: pd-client/client_test.go [hunk 6]
```diff
@@ -175,16 +186,65 @@ func (s *testClientSuite) TestGetRegion(c *C) {
...
+		regions = append(regions, r)
+		req := &pdpb.RegionHeartbeatRequest{
+			Header: newHeader(s.srv),
+			Region: r,
+			Leader: peer,
+		}
+		err := s.regionHeartbeat.Send(req)
+		c.Assert(err, IsNil)
+	}
+	time.Sleep(time.Millisecond * 1000)
```
Comment (rleungx) [inline pd-client/client_test.go]: Second?

File: pd-client/client_test.go [hunk 7]
```diff
@@ -175,16 +186,65 @@ func (s *testClientSuite) TestGetRegion(c *C) {
 	c.Assert(leader, DeepEquals, peer)
 }
 
+func (s *testClientSuite) TestGetPrevRegion(c *C) {
```
Comment (nolouch) [inline pd-client/client_test.go]: I think best do this test as a unit test about `searchPrevRegion`. It easy to fail in ci env.

File: server/core/region.go [hunk 8]
```diff
@@ -561,6 +561,15 @@ func (r *RegionsInfo) SearchRegion(regionKey []byte) *RegionInfo {
 	return r.GetRegion(region.GetId())
 }
 
+// SearchPrevRegion searches previews RegionInfo from regionTree
```
Comment (rleungx) [inline server/core/region.go]: previous

File: server/core/region_tree.go [hunk 9]
```diff
@@ -115,6 +115,23 @@ func (t *regionTree) search(regionKey []byte) *metapb.Region {
 	return result.region
 }
 
+// searchPrev returns the previews region of the region where the regionKey is located.
```
Comment (rleungx) [inline server/core/region_tree.go]: previous

File: pd-client/client.go (hunk 1 above)
Comment (liukun4515) [reply pd-client/client.go]: The requestDuration has been removed except TSO.

**Comment by liukun4515** [comment]
The integration test failed again. @nolouch  @disksing

**Comment by liukun4515** [comment]
PTAL @nolouch @rleungx

**Comment by arthurkiller** [comment]
Awesome work, dude!

**Pull request merged**